### PR TITLE
AP_Mission: Flexible waypoint type #3753

### PR DESCRIPTION
**Pull request by EShamaev**
As for: https://github.com/ArduPilot/ardupilot/issues/3753

Two additional parameters to waypoint logic were added.
These parameters are Acceptance radius in meters and pass through the WP.
The definition was present in MAVLINK.
Due to internal structures there is a limit of 255 meters on both parameters.

Smal change was added to MissionPlanner as well to show the labels: https://github.com/ArduPilot/MissionPlanner/pull/1434

### CODE REVIEW

File: libraries/AP_Mission/AP_Mission.cpp [hunk 1]
```diff
@@ -535,19 +535,23 @@ MAV_MISSION_RESULT AP_Mission::mavlink_int_to_mission_cmd(const mavlink_mission_
...
-        // acceptance radius in meters
-        cmd.p1 = packet.param2;
+        // acceptance radius in meters and pass by distance in meters
+        uint16_t acp = packet.param2;           // param 2 is acceptance radius in meters is held in low p1
+        uint16_t passby = packet.param3;        // param 3 is pass by distance in meters is held in high p1
+        cmd.p1 = (passby << 8) | (acp & 0x00FF);
 #else
-        // delay at waypoint in seconds
-        cmd.p1 = packet.param1;                         
+        // delay at waypoint in seconds (this is for copters???)
```
Comment (khancyr) [inline libraries/AP_Mission/AP_Mission.cpp]: and rover !

**Comment by magicrub** [comment]
This will get a little confusing to the user unless we also implement https://github.com/ArduPilot/ardupilot/issues/3483 because there would be two ways to accomplish the same thing without the change.

**Comment by EShamaev** [comment]
What -1 will do? Did not understand at all. How optimum is calculated?

**Comment by magicrub** [comment]
It can figure out what is the best turn given various inputs. Inputs could be, but not limited to, current state (velocity) or user params (roll limit)

**Comment by EShamaev** [comment]
As far as I understand 0 will be automatic as it is now even if in future any better calculation will be implemented and will replace WP_RADIUS method we have now.

**Comment by EShamaev** [comment]
What can be done with good results, is scale WP_RADIUS with speed and wind direction. Other than that, the WP_RADIUS has showed good results.

**Comment by magicrub** [comment]
Sadly, this Parameter actually has no relation to what it's name implies. This is the radius that it starts considering a turn. It does not specify when to turn and it does not have any relation to a size of a waypoint.

**Comment by EShamaev** [comment]
Yes, I know that. It's an underfly distance to WP. And it works reasonably well, but can be scaled with speed and wind.

**Comment by magicrub** [comment]
Except it's not. Setting it to 100 vs 10000 will have the same flight
behavior.

**Comment by EShamaev** [comment]
Should not.
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_L1_Control/AP_L1_Control.cpp#L120

**Comment by EShamaev** [comment]
Got it.

    wp_radius *= sq(_ahrs.get_EAS2TAS());
    return MIN(wp_radius, _L1_dist);

It already uses speed as a factor, which is good, however I would add wind to it as we are considering ground speed in navigation.

And also
_L1_dist = 0.3183099f * _L1_damping * _L1_period * groundSpeed;
Which also seems to be right as these are time specific capabilities of a nav controller.

So it is already automatic near optimal value.
Why would we need WP_RADIUS set to -1?

**Comment by EShamaev** [comment]
Changed the whole realization of feature. Now much more correct behaviour.
image

**Comment by magicrub** [comment]
merged, thanks! I also added some clarity to the variable names and protected from a divide by zero. Great work!

@khancyr please open a new PR for rover support

**Comment by magicrub** [comment]
Mission planner issue to take advantage of this feature: https://github.com/ArduPilot/MissionPlanner/issues/1449

**Pull request closed without merging**